### PR TITLE
MINOR: Refactor .gitignore

### PR DESCRIPTION
**Pull request by granthenke**
Refactor .gitignore with thorough coverage compiled from https://github.com/github/gitignore and inline documentation of why files are ignored.

### CODE REVIEW

**Comment by ijuma** [comment]
Do we really want to have all of this in `.gitignore`? Personally, I'd prefer to only have the ones that are relevant to Kafka.

**Comment by granthenke** [comment]
@ijuma Agreed this has quite a bit of things included. Generally in `.gitignore` files there are 2 categories, `global` and `project`. The `global` ignore settings are usually in your home git config and are related to your specific usage and environment. Things like os files, editor of choice files, etc. While the `project` ignore settings are just for project specific things like java artifacts, logs, etc.

In the case of Kafka, incrementally global ignore settings were slowly being added the the `.gitignore` file to prevent accidental checkins of IDE files, OS files, etc. Instead of slowly adding them one by one as issues were hit, I figured I would just add a relatively comprehensive list of safe ignores settings, global or not.

I also documented what each section of entries ignores putting the Kafka specific rules towards the bottom.

**Comment by ijuma** [comment]
I am not convinced personally (a comprehensive list may even have to be updated more often, in theory), but let's see what the committers say.

**Comment by ewencp** [comment]
@granthenke @ijuma Should we reconsider this? I hate the ginormous list of ignores, but in practice, we can't rely on people setting up their own global gitignore (we know that from practical experience...). Even if we get the core Kafka devs on board, there are plenty of other contributors and its hard to enforce and easy to miss additions of small files in a review, especially binary files where a diff isn't displayed and so the file is easy to pass over by accident. This particular patch might include a bit more than needed and maybe we could prune some of it down, but in practice we end up adding a lot of non-Kafka-specific stuff like IntelliJ files, etc.

**Comment by granthenke** [comment]
@ewencp Happy to re visit this and update it. I will look it over and try to trim it down, let me know if there are any specific entries I should remove.

**Comment by granthenke** [comment]
@ewencp I rebased the patched and pruned some of it down. Let me know if you think I should remove or add anything else.  

Note: #1245 has another incremental change to .gitignore this patch would have avoided

File: .gitignore [hunk 1]
```diff
@@ -1,42 +1,176 @@
...
+GTAGS
+GRTAGS
+GPATH
+cscope.files
+cscope.out
+cscope.in.out
+cscope.po.out
+
+# Covers SVN
+.svn/
```
Comment (ijuma) [inline .gitignore:100]: Do we need this?

**Comment by ijuma** [comment]
A bit of a late response to @ewencp's comment. I still don't like the approach, but if others see the value, then I don't want to block it. I'll leave the review to you. :)

File: .gitignore (hunk 1 above)
Comment (granthenke) [inline .gitignore:100]: It was there before this change. I didn't remove any that already existed.

Comment (ijuma) [inline .gitignore:100]: Ah, probably before we moved to Git.

File: .gitignore [hunk 2]
```diff
@@ -1,42 +1,176 @@
...
+
+# Covers Windows OS
+## Windows image file caches
+Thumbs.db
+ehthumbs.db
+## Folder config file
+Desktop.ini
+## Recycle Bin used on file shares
+$RECYCLE.BIN/
+## Windows Installer files
```
Comment (ewencp) [inline .gitignore:19]: Why would installer files ever be in a Kafka code checkout? I think stuff like this we can trim. The "global" gitignore configs that make sense to include are ones that we have reason to believe would show up commonly in a developer's checkout and aren't covered by other rules (e.g. even if we one day produced a windows installer, it'd end up in `target/` anyway). I'm thinking things like the thumbnail caches, editor backup files, etc.

File: .gitignore [hunk 3]
```diff
@@ -1,42 +1,176 @@
...
+## Windows shortcuts
+*.lnk
+
+# Covers Mac OSX
+.DS_Store
+.AppleDouble
+.LSOverride
+## Thumbnails
+._*
+## Files that might appear in the root of a volume
```
Comment (ewencp) [inline .gitignore:33]: This seems like another set that don't make sense here. How would you get these into your Kafka checkout?

File: .gitignore [hunk 4]
```diff
@@ -1,42 +1,176 @@
...
+GPATH
+cscope.files
+cscope.out
+cscope.in.out
+cscope.po.out
+
+# Covers SVN
+.svn/
+
+# Covers Java
```
Comment (ewencp) [inline .gitignore:102]: See above comment -- these should only show up in directories that are already excluded by other rules.

File: .gitignore [hunk 5]
```diff
@@ -1,42 +1,176 @@
...
+pom.xml.tag
+pom.xml.releaseBackup
+pom.xml.versionsBackup
+pom.xml.next
+release.properties
+dependency-reduced-pom.xml
+buildNumber.properties
+.mvn/timing.properties
+
+# Covers Binary Packages
```
Comment (ewencp) [inline .gitignore:147]: We shouldn't have any of these, right? We could filter them out.

**Comment by ewencp** [comment]
@ijuma I think we can get to a better balance. We already include IDE-specific files, and I think extending that to any files commonly left by editors, file browsers, etc makes sense, but we can still cut down the current version a lot by removing stuff that really has no business being in a Kafka checkout in the first place. I think we can cut another ~50% of the remaining additions to get something that isn't too bloated.

File: .gitignore [hunk 6]
```diff
@@ -1,42 +1,176 @@
...
+## File-based project format:
+*.ipr
+*.iws
+## Plugin-specific files:
+# IntelliJ
+/out/
+## JIRA plugin
+atlassian-ide-plugin.xml
+
+# Ignore tags created by etags, ctags, gtags (GNU global) and cscope
```
Comment (ewencp) [inline .gitignore:83]: Do any of these get generated in a Java project?

**Comment by guozhangwang** [comment]
@granthenke Would you like to continue working on this PR addressing @ewencp 's comments?

**Comment by hachikuji** [comment]
Closing this due to inactivity. Please rebase and reopen if you think it is still useful.

**Pull request closed without merging**